README.md
# collective.handleclient
Plone add-on to interact with a Handle-based PID service


CHANGES.rst
Changelog
=========


0.1 (unreleased)
----------------

- Initial release.
  [raphael-ritz]


src/collective/handleclient/tests/test_crud.py
# -*- coding: utf-8 -*-
"""Setup tests for this package."""
from collective.handleclient.testing \
  import COLLECTIVE_HANDLECLIENT_INTEGRATION_TESTING  # noqa
from zope.annotation import IAnnotations

from plone.app.testing import TEST_USER_ID
from plone.app.testing import TEST_USER_NAME
from plone.app.testing import login
from plone.app.testing import setRoles

import unittest2 as unittest
from mock import Mock

from collective.handleclient.handle_client import HandleError

KEY = 'collective.handleclient.handle'


class TestCRUD(unittest.TestCase):
    """Test the CRUD calls of collective.handleclient"""

    layer = COLLECTIVE_HANDLECLIENT_INTEGRATION_TESTING

    def setUp(self):
        """Custom shared utility setup for tests."""
        self.portal = self.layer['portal']
        setRoles(self.portal, TEST_USER_ID, ['Manager'])
        login(self.portal, TEST_USER_NAME)
        self.portal.invokeFactory('Document', 'page', title=u'Page')
        client = self.portal.handle_client
        request = self.layer['request']
        initial_config = dict(baseurl='http://example.com',
                              username='123',
                              password='secret')
        request.form['config'] = initial_config
        client.manage_setConfiguration(request)
        uid = self.portal.page.UID()
        self.handle = '/'.join([client.prefix, uid])
        self.handle_url = client.baseurl + self.handle

    def test_create(self):
        """Test creation of a handle"""
        client = self.portal.handle_client
        page = self.portal.page

        response = Mock()
        response.status_code = 201
        response.headers = {"location": self.handle_url}

        client.session.post = Mock(return_value=response)
        hdl = client.create(page)

        self.assertEqual(hdl, self.handle_url)

    def test_epic_create(self):
        """Test EPIC's creation of a handle - they require a 'put'"""
        client = self.portal.handle_client
        page = self.portal.page

        response1 = Mock()
        response1.status_code = 405
        response2 = Mock()
        response2.status_code = 201
        response2.headers = {"location": self.handle_url}

        client.session.post = Mock(return_value=response1)
        client.session.put = Mock(return_value=response2)
        hdl = client.create(page)

        self.assertEqual(hdl, self.handle_url)

    def test_failed_create(self):
        """Test that a failed create request raises HandleError"""
        client = self.portal.handle_client
        page = self.portal.page

        response = Mock()
        response.status_code = 403
        response.headers = {"location": self.handle_url}

        client.session.post = Mock(return_value=response)
        message = "403 You don't have the right to do that."
        with self.assertRaisesRegexp(HandleError, message):
            client.create(page)

    def test_read_unregistered(self):
        """Test reading the registration of an unregistered object"""
        client = self.portal.handle_client
        page = self.portal.page

        response = Mock()
        response.status_code = 204
        response.headers = {"location": self.handle_url}

        client.session.get = Mock(return_value=response)
        hdl = client.read(page)

        self.assertEqual(hdl, None)

    def test_read_registered(self):
        """Test reading the registration of a registered object"""
        client = self.portal.handle_client
        page = self.portal.page

        # fake the registration
        annotations = IAnnotations(page)
        annotations[KEY] = self.handle

        response = Mock()
        response.status_code = 204
        response.headers = {"location": self.handle_url}
        response.text = ''

        client.session.get = Mock(return_value=response)
        hdl = client.read(page)

        self.assertEqual(hdl, self.handle_url)

    def test_not_found(self):
        """Test read resulting in 'not found' raises HandleError"""
        client = self.portal.handle_client
        page = self.portal.page

        # fake the registration
        annotations = IAnnotations(page)
        annotations[KEY] = self.handle

        response = Mock()
        response.status_code = 404
        client.session.get = Mock(return_value=response)

        with self.assertRaisesRegexp(HandleError, "404 Handle not found."):
            client.read(page)

    def test_update_unregistered(self):
        """
        Test updating the registration of an unregistered object
        returns None
        """
        client = self.portal.handle_client
        page = self.portal.page

        response = Mock()
        response.status_code = 204
        response.headers = {"location": self.handle_url}

        client.session.put = Mock(return_value=response)
        hdl = client.update(page)

        self.assertEqual(hdl, None)

    def test_update_registered(self):
        """Test updating the registration of a registered object"""
        client = self.portal.handle_client
        page = self.portal.page

        # fake the registration
        annotations = IAnnotations(page)
        annotations[KEY] = self.handle

        response = Mock()
        response.status_code = 204
        response.headers = {"location": self.handle_url}
        response.text = ''

        client.session.put = Mock(return_value=response)
        hdl = client.update(page)

        self.assertEqual(hdl, self.handle_url)

    def test_delete(self):
        """Test unregistering a registered object"""
        client = self.portal.handle_client
        page = self.portal.page

        # fake the registration
        annotations = IAnnotations(page)
        annotations[KEY] = self.handle

        response = Mock()
        response.status_code = 204

        client.session.delete = Mock(return_value=response)
        client.delete(page)

        self.assertTrue(KEY not in annotations)


src/collective/handleclient/tests/test_view.py
# -*- coding: utf-8 -*-
"""Setup tests for this package."""
from collective.handleclient.testing \
  import COLLECTIVE_HANDLECLIENT_INTEGRATION_TESTING  # noqa

from plone.app.testing import TEST_USER_ID
from plone.app.testing import TEST_USER_NAME
from plone.app.testing import login
from plone.app.testing import setRoles

import unittest2 as unittest
from mock import Mock


KEY = 'collective.handleclient.handle'


class TestView(unittest.TestCase):
    """Test the browser view calls of collective.handleclient"""

    layer = COLLECTIVE_HANDLECLIENT_INTEGRATION_TESTING

    def setUp(self):
        """
        Custom shared utility setup for tests.
        """
        self.portal = self.layer['portal']
        setRoles(self.portal, TEST_USER_ID, ['Manager'])
        login(self.portal, TEST_USER_NAME)
        self.portal.invokeFactory('Document', 'page', title=u'Page')
        client = self.portal.handle_client
        request = self.layer['request']
        initial_config = dict(baseurl='http://example.com',
                              username='123',
                              password='secret')
        request.form['config'] = initial_config
        client.manage_setConfiguration(request)
        uid = self.portal.page.UID()
        self.handle = '/'.join([client.prefix, uid])
        self.handle_url = client.baseurl + self.handle

    def test_hasHandle4unregistered(self):
        """
        Test traversal to check for handle of an unregistered object
        """
        view = self.portal.page.restrictedTraverse('hasHandle')
        has_handle = view.hasHandle()
        self.assertFalse(has_handle)
        handle = view.handle()
        self.assertEqual(handle, None)

    def test_registration(self):
        """
        Test handle registration and subsequent checks.
        """
        client = self.portal.handle_client

        response = Mock()
        response.status_code = 201
        response.headers = {"location": self.handle_url}
        client.session.post = Mock(return_value=response)

        view = self.portal.page.restrictedTraverse('hasHandle')

        self.assertEqual(view.create(), self.handle_url)
        self.assertTrue(view.hasHandle())
        self.assertEqual(view.handle(), self.handle)

        response.status_code = 204
        response.text = ''
        client.session.get = Mock(return_value=response)
        client.session.put = Mock(return_value=response)
        client.session.delete = Mock(return_value=response)

        self.assertEqual(view.read(), self.handle_url)
        self.assertEqual(view.update(), self.handle_url)

        self.assertTrue(view.hasDeletePermission())
        self.assertEqual(view.delete(), None)

    def test_redirect(self):
        """
        Test that the redirection flag is honored.
        """
        client = self.portal.handle_client
        page = self.portal.page

        response = Mock()
        response.status_code = 201
        response.headers = {"location": self.handle_url}
        client.session.post = Mock(return_value=response)

        view = self.portal.page.restrictedTraverse('hasHandle')
        url = page.absolute_url() + '/handle_view'
        self.assertEqual(view.create(redirect=True), url)

        response.status_code = 204
        response.text = ''
        client.session.delete = Mock(return_value=response)
        self.assertEqual(view.delete(redirect=True), url)


src/collective/handleclient/tests/test_robot.py
# -*- coding: utf-8 -*-
from collective.handleclient.testing import COLLECTIVE_HANDLECLIENT_ACCEPTANCE_TESTING  # noqa
from plone.app.testing import ROBOT_TEST_LEVEL
from plone.testing import layered

import os
import robotsuite
import unittest


def test_suite():
    suite = unittest.TestSuite()
    current_dir = os.path.abspath(os.path.dirname(__file__))
    robot_dir = os.path.join(current_dir, 'robot')
    robot_tests = [
        os.path.join('robot', doc) for doc in os.listdir(robot_dir)
        if doc.endswith('.robot') and doc.startswith('test_')
    ]
    for robot_test in robot_tests:
        robottestsuite = robotsuite.RobotTestSuite(robot_test)
        robottestsuite.level = ROBOT_TEST_LEVEL
        suite.addTests([
            layered(
                robottestsuite,
                layer=COLLECTIVE_HANDLECLIENT_ACCEPTANCE_TESTING
            ),
        ])
    return suite


src/collective/handleclient/testing.py
# -*- coding: utf-8 -*-
from plone.app.robotframework.testing import REMOTE_LIBRARY_BUNDLE_FIXTURE
from plone.app.testing import applyProfile
from plone.app.testing import FunctionalTesting
from plone.app.testing import IntegrationTesting
from plone.app.testing import PLONE_FIXTURE
from plone.app.testing import PloneSandboxLayer
from plone.testing import z2

import collective.handleclient


class CollectiveHandleclientLayer(PloneSandboxLayer):

    defaultBases = (PLONE_FIXTURE,)

    def setUpZope(self, app, configurationContext):
        self.loadZCML(package=collective.handleclient)

    def setUpPloneSite(self, portal):
        applyProfile(portal, 'collective.handleclient:default')


COLLECTIVE_HANDLECLIENT_FIXTURE = CollectiveHandleclientLayer()


COLLECTIVE_HANDLECLIENT_INTEGRATION_TESTING = IntegrationTesting(
    bases=(COLLECTIVE_HANDLECLIENT_FIXTURE,),
    name='CollectiveHandleclientLayer:IntegrationTesting'
)


COLLECTIVE_HANDLECLIENT_FUNCTIONAL_TESTING = FunctionalTesting(
    bases=(COLLECTIVE_HANDLECLIENT_FIXTURE,),
    name='CollectiveHandleclientLayer:FunctionalTesting'
)


COLLECTIVE_HANDLECLIENT_ACCEPTANCE_TESTING = FunctionalTesting(
    bases=(
        COLLECTIVE_HANDLECLIENT_FIXTURE,
        REMOTE_LIBRARY_BUNDLE_FIXTURE,
        z2.ZSERVER_FIXTURE
    ),
    name='CollectiveHandleclientLayer:AcceptanceTesting'
)


src/collective/handleclient/browser/handle.py
# -*- coding: utf-8 -*-

from zope.annotation import IAnnotations
from AccessControl import getSecurityManager
from Products.Five.browser import BrowserView
from Products.CMFCore.utils import getToolByName


KEY = 'collective.handleclient.handle'  # this should be in config


class HandleView(BrowserView):
    """View methods to deal with handle pids"""

    @property
    def client(self):
        return getToolByName(self.context, 'handle_client')

    def hasHandle(self):
        """
        Returns True if the context is registered.
        False otherwise
        """
        if self.handle() is None:
            return False
        return True

    def handle(self):
        """
        Looks up the handle in the context's annotation.
        Returns None if not found.
        """
        annotations = IAnnotations(self.context)
        return annotations.get(KEY, None)

    def create(self, redirect=False):
        """
        Register the object's UID as PID in the handle system
        Returns the registered location on success
        Raises HandleError on failure
        """
        location = self.client.create(self.context)
        if not redirect:
            return location
        baseurl = self.context.absolute_url()
        return self.request.response.redirect(baseurl + "/handle_view")

    def read(self):
        """
        Gets the current registration from the server
        Returns the registered location on success
        Raises HandleError on failure
        """
        return self.client.read(self.context)

    def update(self):
        """
        Updates the existing registration with the current
        target URL.
        Returns the registered location on success.
        Raises HandleError on failure
        """
        return self.client.update(self.context)

    def delete(self, redirect=False):
        """
        Deletes the existing registration
        Returns None on success.
        Raises HandleError on failure
        """
        result = self.client.delete(self.context)
        if not redirect:
            return result
        baseurl = self.context.absolute_url()
        return self.request.response.redirect(baseurl + "/handle_view")

    def hasDeletePermission(self):
        """
        Helper method to determine whether the current user
        is allowed to unregister a handle.
        """
        sm = getSecurityManager()
        return sm.checkPermission("Handle Client: Delete Handle",
                                  self.context)


src/collective/handleclient/handle_client.py
# -*- coding: utf-8 -*-
"""HandleClient's main class"""

import os
import logging
import requests

# Zope stuff
from zope.annotation import IAnnotations
from Globals import InitializeClass
from AccessControl import ClassSecurityInfo
from OFS.SimpleItem import SimpleItem

# CMF stuff
from Products.CMFCore.permissions import ManagePortal
from Products.CMFCore.utils import UniqueObject

KEY = 'collective.handleclient.handle'


class HandleClient(UniqueObject, SimpleItem):
    """
    Tool for managing handle pids.
    """

    id = 'handle_client'
    meta_type = 'Handle Client'
    # class defaults
    baseurl = ''
    username = ''
    prefix = ''
    password = ''

    security = ClassSecurityInfo()

    def create(self, context):
        """
        Request a handle to be created for the content object
        passed in as context. The handle will be the configured prefix
        plus the objects uid. The absolut url is supplied as target url.
        On success the handle will be stored in the context's annotations.
        """
        uid = context.UID()
        target = context.absolute_url()
        handle = '/'.join([self.prefix, uid])
        data = [{'type': 'URL', 'parsed_data': target}]

        resp = self.session.post(
            self.baseurl + handle, json=data
            )
        if int(resp.status_code) == 405:  # EPIC doesn't support POST
            resp = self.session.put(
                self.baseurl + handle, json=data
                )
        if int(resp.status_code) == 201:
            location = resp.headers.get('location')
            annotations = IAnnotations(context)
            annotations[KEY] = handle
            return location
        else:
            logger = logging.getLogger('collective.handle')
            message = self.baseurl + handle
            message += '\n' + str(resp) + '\n'
            for k, v in resp.headers.items():
                message += "%s = %s\n" % (k, v)
            logger.error(message)
            raise HandleError(resp.status_code)

    def read(self, context):
        """
        Returns the 'location' if context is registered, None otherwise.
        Raises HandleError if the call fails.
        """
        handle = self._getHandle(context)
        if handle is None:
            return None
        resp = self.session.get(self.baseurl + handle)
        if resp.status_code in [200, 204]:
            return resp.headers.get("location", "") + resp.text
        raise HandleError(resp.status_code)

    def update(self, context):
        """
        Change the target URL that the handle resolves to.
        Returns None if context is not registered.
        Raises HandleError if the call fails.

        Should be called when context was moved or renamed
        (aka whenever its URL changed).
        """
        handle = self._getHandle(context)
        if handle is None:
            return None
        target = context.absolute_url()
        data = [{'type': 'URL', 'parsed_data': target}]
        resp = self.session.put(
            self.baseurl + handle, json=data
            )
        if int(resp.status_code) == 204:
            return resp.headers.get("location")
        raise HandleError(resp.status_code)

    def delete(self, context):
        """
        Unregister the handle and remove it from the annotation.
        Returns None if context is not registered.
        Raises HandleError if the call fails.
        """
        handle = self._getHandle(context)
        if handle is None:
            return None
        resp = self.session.delete(self.baseurl + handle)
        if int(resp.status_code) != 204:
            raise HandleError(resp.status_code)
        self._removeHandle(context)

    def _getHandle(self, context):
        """
        Helper method looking up the handle in the context's annotation.
        Returns None if not existing.
        """
        annotations = IAnnotations(context)
        return annotations.get(KEY, None)

    def _removeHandle(self, context):
        annotations = IAnnotations(context)
        del annotations[KEY]

    # for the tool configuration

    security.declareProtected(ManagePortal, 'getConfiguration')
    def getConfiguration(self):
        """returns the config (without password) for the configlet"""

        config = {'baseurl': self.baseurl,
                  'username': self.username,
                  'prefix': self.prefix,
                  }
        return config

    security.declareProtected(ManagePortal, 'manage_setConfiguration')
    def manage_setConfiguration(self, request):
        """setting the configuration from data recieved through a form"""
        data = request.form.copy()
        config = data.get('config', None)
        if config is not None:
            self._setConfiguration(config)

    def _setConfiguration(self, config):
        """Update configuration settings from form data"""

        baseurl = config.get('baseurl')
        # make sure the URL ends with exactly one slash
        self.baseurl = baseurl.rstrip('/') + '/'

        username = config.get('username', None)
        prefix = config.get('prefix', None)
        # only one needs to be given as username and prefix are
        # usually the same in the handle system
        self.username = username or prefix
        self.prefix = prefix or username

        self.password = config.get('password')
        self._updateSession()

    def _updateSession(self, verify_ssl=False):
        self.session = requests.Session()
        if not self.password:
            password = os.environ['HANDLE_PASSWORD']
        else:
            password = self.password
        self.session.auth = (self.username, password)
        self.session.verify = verify_ssl
        self.session.headers.update({
            'Cache-Control': 'no-cache',
            'User-Agent': 'collective.handleclient',
            })


InitializeClass(HandleClient)


class HandleError(Exception):

    def __init__(self, status):

        messages = {
            400: "400 Bad request.",
            401: "401 Authentication failed.",
            403: "403 You don't have the right to do that.",
            404: "404 Handle not found.",
            405: "405 Method Not Allowed.",
            409: "409 Handle already exists.",
            415: "415 Unsupported Media Type",
            'default': "Handle Server error. ",
            }

        status = int(status)
        message = messages.get(status, None)
        self.message = message or messages.get('default') + str(status)

    def __unicode__(self):
        return unicode(self.message)

    def __str__(self):
        return self.message

    def __repr__(self):
        return self.__str__()
